File: packages/crawler-py/shroodler/sitemap.py
from __future__ import annotations

import xml.etree.ElementTree as ET


def parse_robots_sitemaps(body: str) -> list[str]:
    """Return Sitemap: URLs from a robots.txt body. Never raises."""
    out: list[str] = []
    if not body:
        return out
    for raw in body.splitlines():
        line = raw.strip()
        if not line or line.startswith("#"):
            continue
        lower = line.lower()
        if not lower.startswith("sitemap:"):
            continue
        rest = line.split(":", 1)[1].strip()
        if not rest:
            continue
        token = rest.split()[0]
        if token.startswith("#"):
            continue
        out.append(token)
    return out


def _local_name(tag: str) -> str:
    if tag.startswith("{") and "}" in tag:
        return tag.rsplit("}", 1)[-1].lower()
    return tag.lower()


def parse_sitemap_xml(body: str) -> tuple[list[str], list[str]]:
    """Parse urlset/sitemapindex XML.

    Returns (url_locs, nested_sitemap_locs). Malformed XML returns empty lists.
    """
    url_locs: list[str] = []
    sitemap_locs: list[str] = []
    if not body or not body.strip():
        return url_locs, sitemap_locs
    try:
        root = ET.fromstring(body)
    except ET.ParseError:
        return url_locs, sitemap_locs
    except Exception:
        return url_locs, sitemap_locs
    _walk_sitemap(root, url_locs, sitemap_locs)
    return url_locs, sitemap_locs


def _walk_sitemap(parent: ET.Element, url_locs: list[str], sitemap_locs: list[str]) -> None:
    parent_name = _local_name(parent.tag)
    for child in list(parent):
        name = _local_name(child.tag)
        if name == "loc":
            text = (child.text or "").strip()
            if not text:
                continue
            if parent_name == "sitemap":
                sitemap_locs.append(text)
            else:
                url_locs.append(text)
            continue
        _walk_sitemap(child, url_locs, sitemap_locs)

File: packages/crawler-go/internal/extractors/sitemap.go
package extractors

import (
	"encoding/xml"
	"io"
	"strings"
)

// ParseRobotsSitemaps returns Sitemap: directive URLs from a robots.txt body.
func ParseRobotsSitemaps(body string) []string {
	var out []string
	for _, raw := range strings.Split(body, "\n") {
		line := strings.TrimSpace(raw)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		lower := strings.ToLower(line)
		if !strings.HasPrefix(lower, "sitemap:") {
			continue
		}
		rest := strings.TrimSpace(line[len("Sitemap:"):])
		if i := strings.Index(lower, "sitemap:"); i >= 0 {
			rest = strings.TrimSpace(line[i+len("sitemap:"):])
		}
		if rest == "" {
			continue
		}
		token := strings.Fields(rest)[0]
		if strings.HasPrefix(token, "#") {
			continue
		}
		out = append(out, token)
	}
	return out
}

// ParseSitemapXML returns url <loc> values and nested sitemap-index <loc> values.
// Malformed XML does not panic; it returns whatever was parsed before the error.
func ParseSitemapXML(body string) (urlLocs, sitemapLocs []string) {
	if strings.TrimSpace(body) == "" {
		return nil, nil
	}
	dec := xml.NewDecoder(strings.NewReader(body))
	dec.CharsetReader = func(_ string, input io.Reader) (io.Reader, error) {
		return input, nil
	}
	var stack []string
	for {
		tok, err := dec.Token()
		if err != nil {
			break
		}
		switch t := tok.(type) {
		case xml.StartElement:
			local := strings.ToLower(t.Name.Local)
			stack = append(stack, local)
			if local != "loc" {
				continue
			}
			var loc string
			if err := dec.DecodeElement(&loc, &t); err != nil {
				if len(stack) > 0 {
					stack = stack[:len(stack)-1]
				}
				continue
			}
			loc = strings.TrimSpace(loc)
			parent := ""
			if len(stack) >= 2 {
				parent = stack[len(stack)-2]
			}
			if loc != "" {
				if parent == "sitemap" {
					sitemapLocs = append(sitemapLocs, loc)
				} else {
					urlLocs = append(urlLocs, loc)
				}
			}
			if len(stack) > 0 {
				stack = stack[:len(stack)-1]
			}
		case xml.EndElement:
			if len(stack) == 0 {
				continue
			}
			stack = stack[:len(stack)-1]
		}
	}
	return urlLocs, sitemapLocs
}
